lib/datadog/di/probe_manager.rb

```diff
@@ -0,0 +1,182 @@
+            logger.warn("Unhandled exception in definition trace point: #{exc.class}: #{exc}")
+            telemetry&.report(exc, description: "Unhandled exception in definition trace point")
+            # TODO test this path
+          end
+        end
+      end
+
+      attr_reader :logger
+      attr_reader :telemetry
+
+      # TODO test that close is called during component teardown and
+      # the trace point is cleared
+      def close
+        definition_trace_point.disable
+        clear_hooks
+      end
+
+      def clear_hooks
+        pending_probes.clear
+        installed_probes.each do |probe_id, probe|
+          instrumenter.unhook(probe)
+        end
+        installed_probes.clear
+      end
+
+      attr_reader :settings
+      attr_reader :instrumenter
+      attr_reader :probe_notification_builder
+      attr_reader :probe_notifier_worker
+      attr_reader :installed_probes
+      attr_reader :pending_probes
+
+      # Probes that failed to instrument for reasons other than the target is
+      # not yet loaded are added to this collection, so that we do not try
+      # to instrument them every time remote configuration is processed.
+      attr_reader :failed_probes
+
+      # config is one probe info
+      def add_probe(probe)
+        # TODO lock here?
+
+        # Probe failed to install previously, do not try to install it again.
+        if msg = failed_probes[probe.id]
+          # TODO test this path
+          raise Error::ProbePreviouslyFailed, msg
+        end
+
+        begin
+          instrumenter.hook(probe, &method(:probe_executed_callback))
+
+          installed_probes[probe.id] = probe
+          payload = probe_notification_builder.build_installed(probe)
+          probe_notifier_worker.add_status(payload)
+          # The probe would only be in the pending probes list if it was
+          # previously attempted to be installed and the target was not loaded.
+          # Always remove from pending list here because it makes the
+          # API smaller and shouldn't cause any actual problems.
+          pending_probes.delete(probe.id)
+          true
+        rescue Error::DITargetNotDefined => exc
+          pending_probes[probe.id] = probe
+          false
+        end
+      rescue => exc
+        # In "propagate all exceptions" mode we will try to instrument again.
+        raise if settings.dynamic_instrumentation.internal.propagate_all_exceptions
+
+        logger.warn("Error processing probe configuration: #{exc.class}: #{exc}")
+        telemetry&.report(exc, description: "Error processing probe configuration")
+        # TODO report probe as failed to agent since we won't attempt to
+        # install it again.
+
+        # TODO add top stack frame to message
+        failed_probes[probe.id] = "#{exc.class}: #{exc}"
+
+        raise
+      end
+
+      def remove_other_probes(probe_ids)
+        pending_probes.values.each do |probe|
```

Comment:
```suggestion
        pending_probes.each_value do |probe|
```

Reply:
The way I had it was intentional because I am modifying the container during iteration.